draft-ietf-dnsop-dns-catalog-zones.md

```diff
@@ -220,7 +220,7 @@ More properties may be defined in future documents.
 ### Schema Version (`version` property)
 
 The catalog zone schema version is specified by an integer value embedded in a TXT RR named `version.$CATZ`.
-All catalog zones MUST have a TXT RRset named `version.$CATZ` with at least one RR.
+All catalog zones MUST have a TXT RRset named `version.$CATZ` with exactly one RR.
```

Comment:
The "at least one RR" was to allow for backwards compatibility. For example version 3 introduces some new properties that can be safely ignored by consumers of catalogs of version 2. The producer can in that case put in two RRs in the set, one for 2 and one for 3.

Reply:
> The "at least one RR" was to allow for backwards compatibility. For example version 3 introduces some new properties that can be safely ignored by consumers of catalogs of version 2. The producer can in that case put in two RRs in the set, one for 2 and one for 3.

When the consumer can safely ignore the new property, there is no need to bump the the version number. Version changes are only necessary for non backwards compatible changes. Because of this you never require more than one version record in a zone.

Reply:
I agree that just adding new properties does not require bumping new version number. However, what if some (now unforeseen) changes to catalog zone structure will eventually introduce incompatible change? In that case, we would need to bump the version string number to "3" or more. And it that situation, it might still be useful to generate catalogs that can be interpreted both as "3" and "2". I support keeping the possiblity of multiple version strings, that means the "at least" phrase in this case.

Reply:
While I can't prove a negative, I find it hard to imagine a backwards incompatible change where we could still generate a catalog that works for both versions. I'm open to leaving room for "both" but I still prefer "one version".

Reply:
I find it hard to imagine that too, but to be on the safe side it would be better to maintain "at least" as it also allows to specify a version 3 that is not compatible with version 2 (by having a single version RR). But not the other way around. An "exactly one" does not allow for catalog producers to generate catalogs suitable for version 2 and 3.

Reply:
Thought about it some more - such a mixed-version catalog would necessarily be so complex that nobody should be writing the code for it, producer or consumer. So, let's please stick to one version number in a catalog. This makes life simpler for everybody.

Reply:
Are you sure? A consumer just supports 1 version and looks for it in the RRset.
A producer can either produce for multiple version, or for a single version.

Reply:
(Even the consumer might be able to consume multiple versions of catalog if needed.)

What if the incompatible change, which will eventually require bumping the version number, would only relate to some (marginal) catalog features? In that case, simple catalog zones might be version-universal and deserving multi-version.

I don't see whom would restriction to single version string would make "life simpler". Catalog generator can add version fields to its liking (just at least one), interpret can simply search for any known version listed. This is not difficult to implement for anyone.

Reply:
> What if the incompatible change, which will eventually require bumping the version number, would only relate to some (marginal) catalog features? In that case, simple catalog zones might be version-universal and deserving multi-version.

Again, I have trouble imagining a situation fitting this description. If it's incompatible, you can't leave "2" in there. If it's marginal, it doesn't need "3".

However, we are trying to predict the future here. My prediction is that a single number will be simpler in the future. Your prediction is that supporting a range will be simpler in the future. We don't know who will be right :)

Reply:
Anyway, both options are possible in the future anyway. We can always in an standard defining a new "backwards compatible" catalog zone specify that it's version is specified in a new property `newversion.$CATZ TXT "1"` ;)

Reply:
That is an excellent point, that I'd like to hijack to argue for a single version RR today.

Reply:
I surrender and accept also forcing a single version string. In that case, the rest of the paragrph needs re-phrasing, too.

Reply:
Ok won't change the already committed change, but alter the rest of the paragraph